From da51271d4a27a06d37b87139edd12dd2cf506273 Mon Sep 17 00:00:00 2001
From: Lu Dai <lu.dai@mind.be>
Date: Thu, 27 Feb 2025 12:23:59 +0100
Subject: [PATCH 1/2] controller: detect existence of other controller

misconfiguration of multiple controllers in one network leads to confusing trouble.
this commit adds a check in controller's `topology response` handling, iterate the
`supportedService` TLV, print error log when any AL device claims it is a controller.

Signed-off-by: Lu Dai <lu.dai@mind.be>
---
 .../beerocks/master/tasks/topology_task.cpp    | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/controller/src/beerocks/master/tasks/topology_task.cpp b/controller/src/beerocks/master/tasks/topology_task.cpp
index cd44c3752f..e6025020e4 100644
--- a/controller/src/beerocks/master/tasks/topology_task.cpp
+++ b/controller/src/beerocks/master/tasks/topology_task.cpp
@@ -25,6 +25,7 @@
 #include <tlvf/wfa_map/tlvApOperationalBSS.h>
 #include <tlvf/wfa_map/tlvClientAssociationEvent.h>
 #include <tlvf/wfa_map/tlvProfile2ReasonCode.h>
+#include <tlvf/wfa_map/tlvSupportedService.h>
 
 #include <beerocks/tlvf/beerocks_message_1905_vs.h>
 
@@ -75,6 +76,23 @@ bool topology_task::handle_topology_response(const sMacAddr &src_mac,
         return false;
     }
 
+    auto tlvSupportedService = cmdu_rx.getClass<wfa_map::tlvSupportedService>();
+    if (tlvSupportedService) {
+        auto num_service = tlvSupportedService->supported_service_list_length();
+        for (decltype(num_service) i = 0; i < num_service; i++) {
+            bool ret;
+            wfa_map::tlvSupportedService::eSupportedService val;
+            std::tie(ret, val) =
+                tlvSupportedService->supported_service_list(static_cast<size_t>(i));
+            if (ret == false)
+                break;
+            if (val == wfa_map::tlvSupportedService::eSupportedService::MULTI_AP_CONTROLLER ||
+                val == wfa_map::tlvSupportedService::eSupportedService::EM_AP_CONTROLLER) {
+                LOG(ERROR) << "Dectected existence of another controller in the network";
+            }
+        }
+    }
+
     const auto &al_mac = tlvDeviceInformation->mac();
 
     auto agent = database.m_agents.get(al_mac);
-- 
GitLab


From b0e2a61fc60e625cbd008861edb3868694cb97b6 Mon Sep 17 00:00:00 2001
From: Richard Yu <richard.yu@mitrastar.com.tw>
Date: Tue, 13 May 2025 22:35:57 +0800
Subject: [PATCH 2/2] controller and agent: enhance logging for detection of
 multiple controllers

This commit refines seven existing log messages to improve the
diagnosis of scenarios where multiple controllers are present on the
network, a situation that may arise due to misconfigurations. The
enhancements aim to assist in identifying the presence of unexpected
controllers by providing clearer and more consistent log messages.

Key updates:
1. Implemented a check to prevent the controller from erroneously
   logging its own topology response as originating from another
   controller. Additionally, limited the logging to a maximum of one
   entry per received topology response.
2. Standardized the log messages by setting their level to ERROR and
   prepending them with the [Multiple Controllers Detected] tag. This
   uniformity facilitates easier identification and filtering of
   relevant log entries during debugging sessions.

Refined log messages include:
LOG(ERROR) << "[Multiple Controllers Detected] Received TOPOLOGY_RESPONSE_MESSAGE from another Controller: " << al_mac;
LOG(ERROR) << "[Multiple Controllers Detected] This agent has a local controller with mac=" << db->bridge.mac << " but response came from src_mac=" << src_mac << ", ignoring";
LOG(ERROR) << "[Multiple Controllers Detected] current controller_bridge_mac=" << db->controller_info.bridge_mac << " but response came from src_mac=" << src_mac << ", ignoring";
LOG(ERROR) << "[Multiple Controllers Detected] Ignoring AP-Autoconfiguration Renew Message from an unknown Controller: " << src_mac;
LOG(ERROR) << "[Multiple Controllers Detected] Ignoring CLIENT_CAPABILITY_QUERY_MESSAGE from an unknown Controller: " << src_mac;
LOG(ERROR) << "[Multiple Controllers Detected] Ignoring AP_CAPABILITY_QUERY_MESSAGE from an unknown Controller: " << src_mac;
LOG(ERROR) << "[Multiple Controllers Detected] Ignoring BACKHAUL_STA_CAPABILITY_QUERY_MESSAGE from an unknown Controller: " << src_mac;

https://prplfoundationcloud.atlassian.net/browse/PPM-2373

Signed-off-by: Richard Yu <richard.yu@mitrastar.com.tw>
---
 .../slave/tasks/ap_autoconfiguration_task.cpp      | 14 +++++++++-----
 .../slave/tasks/capability_reporting_task.cpp      |  9 ++++++---
 .../src/beerocks/master/tasks/topology_task.cpp    | 14 ++++++++++----
 3 files changed, 25 insertions(+), 12 deletions(-)
 mode change 100644 => 100755 agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
 mode change 100644 => 100755 agent/src/beerocks/slave/tasks/capability_reporting_task.cpp
 mode change 100644 => 100755 controller/src/beerocks/master/tasks/topology_task.cpp

diff --git a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
old mode 100644
new mode 100755
index 25d631005d..81a790a724
--- a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
+++ b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
@@ -962,14 +962,16 @@ void ApAutoConfigurationTask::handle_ap_autoconfiguration_response(
 {
     auto db = AgentDB::get();
     if (db->device_conf.local_controller && src_mac != db->bridge.mac) {
-        LOG(INFO) << "This agent has a local controller with mac=" << db->bridge.mac
-                  << " but response came from src_mac=" << src_mac << ", ignoring";
+        LOG(ERROR) << "[Multiple Controllers Detected] This agent has a local controller with mac="
+                   << db->bridge.mac << " but response came from src_mac=" << src_mac
+                   << ", ignoring";
         return;
     }
     if (db->controller_info.bridge_mac != network_utils::ZERO_MAC &&
         src_mac != db->controller_info.bridge_mac) {
-        LOG(INFO) << "current controller_bridge_mac=" << db->controller_info.bridge_mac
-                  << " but response came from src_mac=" << src_mac << ", ignoring";
+        LOG(ERROR) << "[Multiple Controllers Detected] current controller_bridge_mac="
+                   << db->controller_info.bridge_mac
+                   << " but response came from src_mac=" << src_mac << ", ignoring";
         return;
     }
 
@@ -1303,7 +1305,9 @@ void ApAutoConfigurationTask::handle_ap_autoconfiguration_wsc_renew(
     LOG(DEBUG) << "AP-Autoconfiguration Renew Message from Controller " << src_mac;
     auto db = AgentDB::get();
     if (src_mac != db->controller_info.bridge_mac) {
-        LOG(ERROR) << "Ignoring AP-Autoconfiguration Renew Message from an unknown Controller";
+        LOG(ERROR) << "[Multiple Controllers Detected] Ignoring AP-Autoconfiguration Renew Message "
+                      "from an unknown Controller: "
+                   << src_mac;
         return;
     }
 
diff --git a/agent/src/beerocks/slave/tasks/capability_reporting_task.cpp b/agent/src/beerocks/slave/tasks/capability_reporting_task.cpp
old mode 100644
new mode 100755
index 9689c33b9d..9725132217
--- a/agent/src/beerocks/slave/tasks/capability_reporting_task.cpp
+++ b/agent/src/beerocks/slave/tasks/capability_reporting_task.cpp
@@ -94,7 +94,8 @@ void CapabilityReportingTask::handle_client_capability_query(ieee1905_1::CmduMes
 
     auto db = AgentDB::get();
     if (src_mac != db->controller_info.bridge_mac) {
-        LOG(ERROR) << "Ignoring CLIENT_CAPABILITY_QUERY_MESSAGE from an unknown Controller: "
+        LOG(ERROR) << "[Multiple Controllers Detected] Ignoring CLIENT_CAPABILITY_QUERY_MESSAGE "
+                      "from an unknown Controller: "
                    << src_mac;
         return;
     }
@@ -464,7 +465,8 @@ void CapabilityReportingTask::handle_ap_capability_query(ieee1905_1::CmduMessage
 
     auto db = AgentDB::get();
     if (src_mac != db->controller_info.bridge_mac) {
-        LOG(ERROR) << "Ignoring AP_CAPABILITY_QUERY_MESSAGE from an unknown Controller: "
+        LOG(ERROR) << "[Multiple Controllers Detected] Ignoring AP_CAPABILITY_QUERY_MESSAGE from "
+                      "an unknown Controller: "
                    << src_mac;
         return;
     }
@@ -644,7 +646,8 @@ void CapabilityReportingTask::handle_backhaul_sta_capability_query(
 
     auto db = AgentDB::get();
     if (src_mac != db->controller_info.bridge_mac) {
-        LOG(ERROR) << "Ignoring BACKHAUL_STA_CAPABILITY_QUERY_MESSAGE from an unknown Controller: "
+        LOG(ERROR) << "[Multiple Controllers Detected] Ignoring "
+                      "BACKHAUL_STA_CAPABILITY_QUERY_MESSAGE from an unknown Controller: "
                    << src_mac;
         return;
     }
diff --git a/controller/src/beerocks/master/tasks/topology_task.cpp b/controller/src/beerocks/master/tasks/topology_task.cpp
old mode 100644
new mode 100755
index e6025020e4..8aa54aaf18
--- a/controller/src/beerocks/master/tasks/topology_task.cpp
+++ b/controller/src/beerocks/master/tasks/topology_task.cpp
@@ -76,7 +76,9 @@ bool topology_task::handle_topology_response(const sMacAddr &src_mac,
         return false;
     }
 
-    auto tlvSupportedService = cmdu_rx.getClass<wfa_map::tlvSupportedService>();
+    const auto &al_mac                = tlvDeviceInformation->mac();
+    bool supported_controller_service = false;
+    auto tlvSupportedService          = cmdu_rx.getClass<wfa_map::tlvSupportedService>();
     if (tlvSupportedService) {
         auto num_service = tlvSupportedService->supported_service_list_length();
         for (decltype(num_service) i = 0; i < num_service; i++) {
@@ -88,12 +90,16 @@ bool topology_task::handle_topology_response(const sMacAddr &src_mac,
                 break;
             if (val == wfa_map::tlvSupportedService::eSupportedService::MULTI_AP_CONTROLLER ||
                 val == wfa_map::tlvSupportedService::eSupportedService::EM_AP_CONTROLLER) {
-                LOG(ERROR) << "Dectected existence of another controller in the network";
+                supported_controller_service = true;
+                break;
             }
         }
     }
-
-    const auto &al_mac = tlvDeviceInformation->mac();
+    if (supported_controller_service && al_mac != database.get_local_bridge_mac()) {
+        LOG(ERROR) << "[Multiple Controllers Detected] Received TOPOLOGY_RESPONSE_MESSAGE from "
+                      "another Controller: "
+                   << al_mac;
+    }
 
     auto agent = database.m_agents.get(al_mac);
     if (!agent) {
-- 
GitLab

